### DISCUSSION TITLE
Why does neovide run my shell on startup?

### DISCUSSION THREAD

**Discussion by unphased** [neovide/neovide#2763] [Q&A] [answered]
I'm testing neovide out, because I saw that I can get really beautiful smooth scrolling in neovim with neovide.

on macOS I am initially met with this issue: 

```
Neovide just crashed :(
This is the error that caused the crash. In case you don't know what to do with this, please feel free to report this on https://github.com/neovide/neovide/issues!

ERROR: Unexpected output from neovim binary:
	Hi from /Users/slu/.zprofile
/opt/homebrew/bin/nvim -v
stdout: Hi from /Users/slu/.zprofile
NVIM v0.10.0-dev-3100+gd3fa88b70-Homebrew
Build type: Release
LuaJIT 2.1.1713773202
Run "nvim -V1 -v" for more info

stderr: zsh:1: command not found: Hi
	

Please check your shell configuration.
$SHELL -lc '{bin} -v'

```

This makes no sense for a number of reasons but I can see that it is choking on some basic echo output from my `~/.zprofile` config. My question .... well, i'm at a loss as to where to even start. What's going on here?

- To even get the output "Hi from" zsh must be launched, otherwise why would ~/.zprofile be referenced. Hence, why is zsh being launched?
- Where is the nvim version info coming from? nvim is also being launched. 
- why is zsh interactive shell's initial line of stdout output being EXECUTED BY ZSH?

**Comment by unphased** [comment]
neovide's troubleshooting page offers a glimmer of insight,

> Neovide uses your shell to find nvim and can't know the difference between output and nvim's path. 

I've certainly got some concerns with THAT being a reasonable approach, but it gives me a hint as to what's going on here.

**Reply by unphased** [reply]
able now to start neovide from a command line via 

    neovim-bin = "/opt/homebrew/bin/nvim"

inside `~/.config/neovide/config.toml`.

**Reply by unphased** [reply, marked as answer]
That wasn't the solution to the issue though. the real solution is to clear out any echo to stdout from ~/.zprofile when zsh is set to the system shell. there is no way to avoid neovide from trying to run system shell to discover stuff, so in the end I just removed my "echo Hi from zprofile" from my zprofile and I have no more neovide startup woes.

**Comment by fredizzimo** [comment]
The reason why Neovide launches a shell is that many tools used together with Neovim assumes that you are in a shell environment, otherwise things like LSPs or NPM won't work. 

On macOS a login shell is not started together with the desktop environment, like it is on Linux for example, so it's up to the terminal emulator, and Neovim in this case to start it.

Neovide does however *not* launch an interactive shell, it launches a non-interactive login shell, which reads `.zprofile`, but not `.zshrc` (which is reserved for interactive shells), with the assumption that no output is written. It would be nice to not have that constraint, but Neovim's standard UI protocol uses stdout for the communication, so any extra output would cause communication errors.

**Reply by unphased** [reply]
Thank you for the clarification. I guess the issue is just about how at some point `zsh -lc 'nvim -v'` is being launched to obtain nvim's version, and if `.zprofile` has extra stdout output, the handling code can't easily filter that out cleanly and just bails. I can see that nvim as launched by neovide is running under a shell (the system configured shell) for the reason you gave above, but the failure is not *directly* because of that aspect of it, as nvim itself would have no issues (??? i'm not sure about this actually..) running under `zsh -lc nvim --embed -p`.

Probably also worth noting that on my system, not only did I need to change my `~/.zprofile` which had a print I left in it from some debugging I did ages ago, but there is really no guarantee that it would contain everything needed by neovide. For example, my neovide launch is done from hammerspoon: `os.execute("pushd " .. path .. "; PATH=$HOME/.cargo/bin:$PATH /opt/homebrew/bin/neovide &")` 

Maybe this is my own doing, I maintain a shared aliases script that is the same that I source from bashrc and zshrc, and I tend to move cli related PATH definitions into this so I can have a good shell environment whether I happen to be in bash or zsh. One of the tradeoffs is that it breaks some assumptions like the ones you have to employ for the way neovide starts up. I have no idea at this point whether cargo upon being installed installs itself into zprofile, but if it did I may have moved it into my script that only adds cargo to the PATH under interactive shells, I guess I might need to reconsider my architecture for this...
